### PR TITLE
Redo extra groups and all and add core

### PR DESCRIPTION
**Pull request by nabobalis**
Todo:

- [x] Add changelog
- [x] ADd what's new entry
- [x] add policy to the dev guide on this 
- [x] Fix this from showing

```bash
The following packages are not installed for the sunpy[docs-gallery] requirement:
  * astroquery
  * jplephem
  * mplcairo
```

- [x] Fix this from being different from the rest

```
You do not have all the required dependencies installed to run the sunpy test suite.

['hvpy', 'jplephem', 'pytest-astropy', 'pytest-mpl', 'pytest-xdist', 'pytest']
```

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -86,18 +86,19 @@ timeseries = [
   "pandas>=1.4.0",
 ]
 visualization = ["matplotlib>=3.5.0", "mpl-animators>=1.0.0"]
-all = ["sunpy[asdf,dask,image,jpeg2000,map,net,timeseries,visualization]"]
+core = ["sunpy[image,map,net,timeseries,visualization]"]
+all = ["sunpy[all,jpeg2000,opencv,spice,scikit-image]"]
```
Comment (Cadair) [inline pyproject.toml:90]:
```suggestion
all = ["sunpy[core,jpeg2000,opencv,spice,scikit-image]"]
```

Comment (nabobalis) [inline pyproject.toml:90]:
```suggestion
all = ["sunpy[core,jpeg2000,opencv,spice,scikit-image]"]
```

File: pyproject.toml [hunk 2]
```diff
@@ -86,18 +86,19 @@ timeseries = [
...
-  "jplephem>=2.14",          # For some coordinates tests
+  "jplephem>=2.14",
   "pytest-astropy>=0.11.0",
   "pytest-mpl>=0.16",
   "pytest-xdist>=3.0.2",
   "pytest>=7.1.0",
 ]
 docs = [
-  "sunpy[all,spice]",           # Importing sunpy.coordinates.spice raises an error if spiceypy is not installed
+  "sunpy[all,asdf]",  # Importing sunpy.coordinates.spice raises an error if spiceypy is not installed
```
Comment (Cadair) [inline pyproject.toml:101]:
```suggestion
  "sunpy[all,asdf]",
```

Comment (nabobalis) [inline pyproject.toml:101]:
```suggestion
  "sunpy[all,asdf]",
```

File: pyproject.toml [hunk 3]
```diff
@@ -111,12 +112,12 @@ docs = [
   "sphinx-hoverxref>=1.3.0",
 ]
 docs-gallery = [
-  "sunpy[docs,opencv]",
+  "sunpy[docs,asdf]",
```
Comment (nabobalis) [inline pyproject.toml:115]:
```suggestion
  "sunpy[docs]",
```

File: pyproject.toml [hunk 4]
```diff
@@ -86,18 +86,19 @@ timeseries = [
   "pandas>=1.4.0",
 ]
 visualization = ["matplotlib>=3.5.0", "mpl-animators>=1.0.0"]
-all = ["sunpy[asdf,dask,image,jpeg2000,map,net,timeseries,visualization]"]
+core = ["sunpy[image,map,net,timeseries,visualization]"]
+all = ["sunpy[core,jpeg2000,opencv,spice,scikit-image]"]
```
Comment (Cadair) [inline pyproject.toml:90]: I think the remaining question is should we keep asdf in all or not.

Comment (nabobalis) [inline pyproject.toml:90]:
```suggestion
all = ["sunpy[core,asdf,jpeg2000,opencv,spice,scikit-image]"]
```

File: pyproject.toml [hunk 5]
```diff
@@ -86,18 +86,19 @@ timeseries = [
...
 tests = [
-  "sunpy[all,opencv,spice]",
+  "sunpy[all,asdf]",
```
Comment (nabobalis) [inline pyproject.toml:92]:
```suggestion
  "sunpy[all]",
```

File: pyproject.toml [hunk 6]
```diff
@@ -86,18 +86,19 @@ timeseries = [
...
-  "jplephem>=2.14",          # For some coordinates tests
+  "jplephem>=2.14",
   "pytest-astropy>=0.11.0",
   "pytest-mpl>=0.16",
   "pytest-xdist>=3.0.2",
   "pytest>=7.1.0",
 ]
 docs = [
-  "sunpy[all,spice]",           # Importing sunpy.coordinates.spice raises an error if spiceypy is not installed
+  "sunpy[all,asdf]",
```
Comment (nabobalis) [inline pyproject.toml:101]:
```suggestion
  "sunpy[all]",
```

File: pyproject.toml [hunk 7]
```diff
@@ -51,11 +51,10 @@ Changelog = "https://docs.sunpy.org/en/stable/whatsnew/changelog.html"
 "Issue Tracker" = "https://github.com/sunpy/sunpy/issues"
 
 [project.optional-dependencies]
-# The list of available extras is also in the installation docs, if you add or remove one please edit it there as well.
+# The list of available extras is also in the installation docs, if you add or remove one please edit it
+# there as well.
```
Comment (nabobalis) [inline pyproject.toml:55]:
```suggestion
# The list of available extras is also in the installation docs,
# if you add or remove one please edit it there as well.
```

File: tox.ini [hunk 8]
```diff
@@ -123,3 +124,12 @@ deps =
...
     python -c "import sunpy; sunpy.self_test()"
+
+
+[testenv:core_deps]
+description = Check the test suite does not fail if only the core dependencies are insatlled.
+deps =
+extras =
+    core
+commands =
+    python -c "import sunpy; sunpy.self_test()"
```
Comment (Cadair) [inline tox.ini:135]: Does this actually run the test suite? Wouldn't we need the test extra for that?

Maybe this isn't for this PR, but actually I think having things not in the core extra auto-skip in the test suite might not be a bad idea, so we can actually run the tests here.

Comment (nabobalis) [reply tox.ini:135]: Well what do you want to happen?

Do we want to be able to run the test suite with some tests that do not error if only the core deps are installed?

We want them to skip tests, raise a message?

File: changelog/7536.breaking.2.rst [hunk 9]
```diff
@@ -0,0 +1,9 @@
+The "all" extra group now will install all optional packages.
+
+This now includes the following packages:
+
+- asdf
+- glmyur
+- opencv
+- scikit-image
+- spicepy
```
Comment (wtbarnes) [inline changelog/7536.breaking.2.rst:9]:
```suggestion
- spiceypy
```

File: changelog/7536.feature.rst [hunk 10]
```diff
@@ -0,0 +1,10 @@
+Added a "core" extra group that does not install any truly optional dependencies.
+It only includes the dependencies that are required to import sunpy or any of its subpackages.
+
+This means it will not install:
+
+- asdf
+- glymur
+- opencv
+- scikit-image
+- spicepy
```
Comment (wtbarnes) [inline changelog/7536.feature.rst:10]:
```suggestion
- spiceypy
```

File: docs/dev_guide/contents/dependencies.rst [hunk 11]
```diff
@@ -20,3 +23,18 @@ Sponsored affiliated packages will support *at least* the sunpy LTS version at t
...
 
 .. _NEP-0029: https://numpy.org/neps/nep-0029-deprecation_policy.html
+
+Extra Packages and Groups
+=========================
+
+sunpy has a number of optional dependencies that are not required to run the core functionality of the package.
+These dependencies are grouped into "extras" that can be installed with the package.
+
+We have to three categories of extras:
```
Comment (wtbarnes) [inline docs/dev_guide/contents/dependencies.rst:33]:
```suggestion
We have three categories of extras:
```

File: changelog/7536.feature.rst [hunk 12]
```diff
@@ -0,0 +1,10 @@
+Added a "core" extra group that does not install any truly optional dependencies.
+It only includes the dependencies that are required to import sunpy or any of its subpackages.
```
Comment (wtbarnes) [inline changelog/7536.feature.rst:2]:
```suggestion
It only includes the dependencies that are required to import sunpy and all subpackages.
```

File: docs/dev_guide/contents/dependencies.rst [hunk 13]
```diff
@@ -20,3 +23,18 @@ Sponsored affiliated packages will support *at least* the sunpy LTS version at t
...
+
+* subpackages : These extra groups are named after each subpackage and list the requirements that enable the core functionality of that subpackage.
+* optional : These extra groups are named after package or feature that they enable. These are not needed to import sunpy or any of its subpackages, but are needed to use some of the functionality.
+  It is expected that subpackages will handle their import manually with a nice import error message if the package is not installed.
+* groupings : These are the groups that are used to install multiple optional dependencies at once.
+  These range from "all" which is every optional dependency to "core" which only the dependencies that are needed to run the core functionality of sunpy.
+  Also includes the "test", "docs" and "dev" groups.
```
Comment (wtbarnes) [inline docs/dev_guide/contents/dependencies.rst:40]:
```suggestion
* subpackages : These extra groups are named after each subpackage and list the requirements that enable the core functionality of that subpackage.
* optional : These extra groups are named after a package or feature they enable. These are not needed to import sunpy or any of its subpackages, but are needed to use some of the functionality.
  Subpackages will emit a clear import error message if the needed package is not installed.
* groupings : These are the groups that are used to install multiple optional dependencies at once.
  These range from "all" which includes every optional dependency to "core" which only includes the dependencies that are needed to run the core functionality of sunpy.
  This also includes the "test", "docs" and "dev" groups.
```

File: docs/whatsnew/6.0.rst [hunk 14]
```diff
@@ -45,6 +45,18 @@ The minimum required versions of these optional dependencies has also been updat
...
+
+The extra groups have been overhauled to make it easier to install sunpy with only the dependencies you need.
+The following extra group have been added:
+
+- "core" - installs only the dependencies that are required to import sunpy or any of its subpackages.
+
+Now the "all" extra group is a meta group that includes all optional dependencies.
+
+We have removed the "dask" extra group and we have removed "scikit-image" from the "image" extra group and added it a new group called "scikit-image".
+
```
Comment (wtbarnes) [inline docs/whatsnew/6.0.rst:59]:
```suggestion
- "core" - installs only the dependencies that are required to import sunpy or any of its subpackages.
- "scikit-image" - installs the `scikit-image` package used for image processing. This dependency is no longer included in the "image" extra group.

Additionally, the "all" extra group is a meta group that includes all optional dependencies.
The "dask" extra group has also been removed.
```

**Review by wtbarnes — APPROVED** [review]
Clean up language in whats new and changelogs, but otherwise looks good to me.

File: docs/whatsnew/6.0.rst [hunk 15]
```diff
@@ -45,6 +45,17 @@ The minimum required versions of these optional dependencies has also been updat
...
+The extra groups have been overhauled to make it easier to install sunpy with only the dependencies you need.
+The following extra group have been added:
+
+- "core" - installs only the dependencies that are required to import sunpy or any of its subpackages.
+- "scikit-image" - installs the `scikit-image` package used for image processing. This dependency is no longer included in the "image" extra group.
+
+Additionally, the "all" extra group is a meta group that includes all optional dependencies.
+The "dask" extra group has also been removed.
 Removal of ``sunpy.database``
```
Comment (nabobalis) [inline docs/whatsnew/6.0.rst:59]:
```suggestion

Removal of ``sunpy.database``
```

File: docs/whatsnew/6.0.rst [hunk 16]
```diff
@@ -45,6 +45,17 @@ The minimum required versions of these optional dependencies has also been updat
...
 - zeep >= 4.1.0
 
+Overhaul of the extra groups
+============================
+
+The extra groups have been overhauled to make it easier to install sunpy with only the dependencies you need.
+The following extra group have been added:
+
+- "core" - installs only the dependencies that are required to import sunpy or any of its subpackages.
+- "scikit-image" - installs the `scikit-image` package used for image processing. This dependency is no longer included in the "image" extra group.
```
Comment (nabobalis) [inline docs/whatsnew/6.0.rst:55]:
```suggestion
- "scikit-image" - installs the ``scikit-image`` package used for image processing. This dependency is no longer included in the "image" extra group.
```

**Pull request merged**